analyzers/src/SonarAnalyzer.CSharp/Rules/FrameworkViewCompiler.cs

```diff
@@ -0,0 +1,339 @@
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+using Microsoft.CodeAnalysis.CodeFixes;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class FrameworkViewCompiler : SonarDiagnosticAnalyzer
+{
+    // TODO: Read this from context.ProjectConfiguration().ProjectPath.
+    // Not sure how to set it up in the UTs.
+    private const string ROOT_DIR = """C:\dev\personal\aspnet\Lol""";
+
+    private ImmutableArray<DiagnosticAnalyzer> Rules;
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+    protected override bool EnableConcurrentExecution => false;
+
+    public FrameworkViewCompiler()
+    {
+        Rules = RuleFinder2.CreateAnalyzers(AnalyzerLanguage.CSharp, false).ToImmutableArray();
+        SupportedDiagnostics = Rules.SelectMany(x => x.SupportedDiagnostics).ToImmutableArray();
+    }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+
+        context.RegisterCompilationAction(
+            c =>
+            {
+                var mightBeUseful = c.ProjectConfiguration();
+                var supportedDiagnostics = SupportedDiagnostics.ToImmutableDictionary(x => x.Id, x => ReportDiagnostic.Warn);
+
+                var dummy = CompileViews(c.Compilation)
+                    .WithOptions(c.Compilation.Options.WithSpecificDiagnosticOptions(supportedDiagnostics))
+                    .WithAnalyzers(Rules);
+
+                var diagnostics = dummy.GetAllDiagnosticsAsync().Result;
+                foreach (var diagnostic in diagnostics.Where(x => supportedDiagnostics.Keys.Contains(x.Id)))
+                {
+                    // TODO: This breaks when raising on out-of-compilation files. Find a replacement, maybe writing directly to the SARIF.
+                    c.ReportIssue(diagnostic);
+                }
+            });
+
+    Compilation CompileViews(Compilation compilation)
+    {
+        FilesToAnalyzeProvider filesProvider = new(Directory.GetFiles(ROOT_DIR, "*.*", SearchOption.AllDirectories));
+        var razorCompiler = new RazorCompiler(ROOT_DIR, filesProvider);
+        var dummyCompilation = compilation;
+
+        foreach (var razorDocument in razorCompiler.CompileAll().Where(x => x.Source.FilePath.Contains("Contact")))
+        {
+            if (razorDocument.GetCSharpDocument()?.GeneratedCode is { } csharpCode)
+            {
+                var fp = dummyCompilation.SyntaxTrees.FirstOrDefault(x => x.FilePath.Contains("Contact")).FilePath;
+
+                var razorTree = CSharpSyntaxTree.ParseText(csharpCode, new CSharpParseOptions(compilation.GetLanguageVersion()))
+                    .WithFilePath(fp);
+                dummyCompilation = dummyCompilation.AddSyntaxTrees(razorTree);
+            }
+        }
+
+        return dummyCompilation;
+    }
+}
+#region RULE_FINDER
+// Copied from the test framework.
+internal static class RuleFinder2
```

Comment:
TLDR: This class uses reflection and fetches all of our rules from the same assembly.